File: node/src/query/engine.js
/**
 * Detect the search mode from a query.
 * Returns "fts" | "wildcard" | "structured"
 */
export function detectSearchMode(filters, search) {
  if (search) return 'basic';
  for (const value of Object.values(filters || {})) {
    if (typeof value === 'string' && (value.includes('*') || value.includes('?'))) {
      return 'wildcard';
    }
  }
  return 'structured';
}

/**
 * Build WHERE clause fragments from filters.
 * Returns an array of SQL condition strings.
 */
function buildWhereClauses(filters) {
  const clauses = [];

  for (const [key, value] of Object.entries(filters || {})) {
    const col = sanitizeColumnName(key);

    if (typeof value === 'string') {
      if (value.includes('*') || value.includes('?')) {
        const likePattern = value.replace(/\*/g, '%').replace(/\?/g, '_');
        clauses.push(`${col} LIKE '${escapeSql(likePattern)}'`);
      } else {
        clauses.push(`${col} = '${escapeSql(value)}'`);
      }
    } else if (typeof value === 'number') {
      clauses.push(`${col} = ${value}`);
    } else if (Array.isArray(value)) {
      const vals = value.map((v) => `'${escapeSql(String(v))}'`).join(', ');
      clauses.push(`${col} IN (${vals})`);
    } else if (typeof value === 'object' && value !== null) {
      if (value.gt !== undefined) clauses.push(`${col} > ${Number(value.gt)}`);
      if (value.gte !== undefined) clauses.push(`${col} >= ${Number(value.gte)}`);
      if (value.lt !== undefined) clauses.push(`${col} < ${Number(value.lt)}`);
      if (value.lte !== undefined) clauses.push(`${col} <= ${Number(value.lte)}`);
    }
  }

  return clauses;
}

function sanitizeColumnName(name) {
  return `"${name.replace(/"/g, '""')}"`;
}

function escapeSql(str) {
  return str.replace(/'/g, "''");
}

function tenantTableName(tenantId) {
  return tenantId.replace(/[^a-zA-Z0-9]/g, '_');
}

function segmentListSql(segments) {
  return segments.map((s) => `'${escapeSql(s)}'`).join(', ');
}

/**
 * Execute a structured or wildcard query on a tenant table.
 */
async function executeDirectQuery(conn, tableName, segments, filters, { limit = 100, offset = 0, startTs, endTs }) {
  const clauses = buildWhereClauses(filters);
  clauses.push(`"_segment" IN (${segmentListSql(segments)})`);
  if (startTs) clauses.push(`"timestamp" >= ${startTs}`);
  if (endTs) clauses.push(`"timestamp" <= ${endTs}`);

  const where = `WHERE ${clauses.join(' AND ')}`;

  const sql = `SELECT * EXCLUDE (_row_id, _segment) FROM ${tableName} ${where} ORDER BY "timestamp" DESC LIMIT ${Number(limit)} OFFSET ${Number(offset)}`;

  const reader = await conn.runAndReadAll(sql);
  return reader.getRowObjectsJson();
}

/**
 * Execute a basic search query using ILIKE on the message column.
 */
async function executeBasicSearchQuery(conn, tableName, segments, filters, search, { limit = 100, offset = 0, startTs, endTs }) {
  const clauses = buildWhereClauses(filters);
  clauses.push(`"_segment" IN (${segmentListSql(segments)})`);
  if (startTs) clauses.push(`"timestamp" >= ${startTs}`);
  if (endTs) clauses.push(`"timestamp" <= ${endTs}`);
  const searchEscaped = escapeSql(search);
  clauses.push(`"message" ILIKE '%${searchEscaped}%'`);
  const where = `WHERE ${clauses.join(' AND ')}`;

  const sql = `SELECT * EXCLUDE (_row_id, _segment) FROM ${tableName} ${where} ORDER BY "timestamp" DESC LIMIT ${Number(limit)} OFFSET ${Number(offset)}`;

  const reader = await conn.runAndReadAll(sql);
  return reader.getRowObjectsJson();
}

/**
 * Main query execution entry point.
 * @param {object} conn - DuckDB connection
 * @param {string} tenant - tenant ID (e.g. "tenant-1")
 * @param {string[]} segments - segment names to query (e.g. ["2025-01-15_00", "2025-01-15_01"])
 * @param {object} options - { filters, search, limit, offset, startTs, endTs }
 */
export async function execute(conn, tenant, segments, { filters, search, limit, offset, startTs, endTs }) {
  const start = performance.now();
  const tableName = tenantTableName(tenant);
  const searchMode = detectSearchMode(filters, search);

  let rows;
  if (searchMode === 'basic') {
    rows = await executeBasicSearchQuery(conn, tableName, segments, filters, search, { limit, offset, startTs, endTs });
  } else {
    rows = await executeDirectQuery(conn, tableName, segments, filters, { limit, offset, startTs, endTs });
  }

  const queryTimeMs = Math.round(performance.now() - start);

  return {
    rows,
    queryTimeMs,
    searchMode,
    filesScanned: segments.length,
    rowsMatched: rows.length,
  };
}

File: go/internal/query/engine.go
package query

import (
	"database/sql"
	"fmt"
	"strconv"
	"strings"
	"time"

	"ducker/internal/model"
	"ducker/internal/tenant"
)

func DetectSearchMode(filters map[string]any, search string) string {
	if search != "" {
		return "basic"
	}
	for _, value := range filters {
		if s, ok := value.(string); ok {
			if strings.Contains(s, "*") || strings.Contains(s, "?") {
				return "wildcard"
			}
		}
	}
	return "structured"
}

func Execute(conn *sql.DB, tenantID string, segments []string, filters map[string]any, search string, limit, offset int, startTS, endTS int64) (model.QueryResult, error) {
	start := time.Now()
	searchMode := DetectSearchMode(filters, search)
	if len(segments) == 0 {
		return model.QueryResult{Rows: []map[string]any{}, QueryTimeMS: 0, SearchMode: searchMode, FilesScanned: 0, RowsMatched: 0}, nil
	}

	table := tenant.TableName(tenantID)
	var (
		rows []map[string]any
		err  error
	)
	if searchMode == "basic" {
		rows, err = executeBasicSearchQuery(conn, table, segments, filters, search, limit, offset, startTS, endTS)
	} else {
		rows, err = executeDirectQuery(conn, table, segments, filters, limit, offset, startTS, endTS)
	}
	if err != nil {
		return model.QueryResult{}, err
	}

	return model.QueryResult{
		Rows:         rows,
		QueryTimeMS:  int64(time.Since(start) / time.Millisecond),
		SearchMode:   searchMode,
		FilesScanned: len(segments),
		RowsMatched:  len(rows),
	}, nil
}

func executeDirectQuery(conn *sql.DB, table string, segments []string, filters map[string]any, limit, offset int, startTS, endTS int64) ([]map[string]any, error) {
	clauses := buildWhereClauses(filters)
	clauses = append(clauses, fmt.Sprintf(`"_segment" IN (%s)`, segmentListSQL(segments)))
	if startTS > 0 {
		clauses = append(clauses, fmt.Sprintf(`"timestamp" >= %d`, startTS))
	}
	if endTS > 0 {
		clauses = append(clauses, fmt.Sprintf(`"timestamp" <= %d`, endTS))
	}
	where := "WHERE " + strings.Join(clauses, " AND ")
	sqlText := fmt.Sprintf(`SELECT * EXCLUDE (_row_id, _segment) FROM %s %s ORDER BY "timestamp" DESC LIMIT %d OFFSET %d`, table, where, limit, offset)
	return runRowMapQuery(conn, sqlText)
}

func executeBasicSearchQuery(conn *sql.DB, table string, segments []string, filters map[string]any, search string, limit, offset int, startTS, endTS int64) ([]map[string]any, error) {
	clauses := buildWhereClauses(filters)
	clauses = append(clauses, fmt.Sprintf(`"_segment" IN (%s)`, segmentListSQL(segments)))
	if startTS > 0 {
		clauses = append(clauses, fmt.Sprintf(`"timestamp" >= %d`, startTS))
	}
	if endTS > 0 {
		clauses = append(clauses, fmt.Sprintf(`"timestamp" <= %d`, endTS))
	}
	searchEscaped := escapeSQL(search)
	clauses = append(clauses, fmt.Sprintf(`"message" ILIKE '%%%s%%'`, searchEscaped))
	where := "WHERE " + strings.Join(clauses, " AND ")
	sqlText := fmt.Sprintf(`SELECT * EXCLUDE (_row_id, _segment) FROM %s %s ORDER BY "timestamp" DESC LIMIT %d OFFSET %d`, table, where, limit, offset)
	return runRowMapQuery(conn, sqlText)
}

func buildWhereClauses(filters map[string]any) []string {
	clauses := make([]string, 0)
	for key, value := range filters {
		col := sanitizeColumnName(key)
		switch v := value.(type) {
		case string:
			if strings.Contains(v, "*") || strings.Contains(v, "?") {
				likePattern := strings.ReplaceAll(strings.ReplaceAll(v, "*", "%"), "?", "_")
				clauses = append(clauses, fmt.Sprintf("%s LIKE '%s'", col, escapeSQL(likePattern)))
			} else {
				clauses = append(clauses, fmt.Sprintf("%s = '%s'", col, escapeSQL(v)))
			}
		case []any:
			vals := make([]string, 0, len(v))
			for _, item := range v {
				vals = append(vals, fmt.Sprintf("'%s'", escapeSQL(fmt.Sprintf("%v", item))))
			}
			clauses = append(clauses, fmt.Sprintf("%s IN (%s)", col, strings.Join(vals, ", ")))
		case map[string]any:
			appendRangeClauses(&clauses, col, v)
		}
	}
	return clauses
}

func appendRangeClauses(clauses *[]string, col string, v map[string]any) {
	if val, ok := v["gt"]; ok {
		*clauses = append(*clauses, fmt.Sprintf("%s > %s", col, numericLiteral(val)))
	}
	if val, ok := v["gte"]; ok {
		*clauses = append(*clauses, fmt.Sprintf("%s >= %s", col, numericLiteral(val)))
	}
	if val, ok := v["lt"]; ok {
		*clauses = append(*clauses, fmt.Sprintf("%s < %s", col, numericLiteral(val)))
	}
	if val, ok := v["lte"]; ok {
		*clauses = append(*clauses, fmt.Sprintf("%s <= %s", col, numericLiteral(val)))
	}
}

func numericLiteral(v any) string {
	switch n := v.(type) {
	case float64:
		return strconv.FormatFloat(n, 'f', -1, 64)
	case float32:
		return strconv.FormatFloat(float64(n), 'f', -1, 64)
	case int:
		return strconv.Itoa(n)
	case int64:
		return strconv.FormatInt(n, 10)
	default:
		return "0"
	}
}

func runRowMapQuery(conn *sql.DB, sqlText string) ([]map[string]any, error) {
	rows, err := conn.Query(sqlText)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	cols, err := rows.Columns()
	if err != nil {
		return nil, err
	}
	result := make([]map[string]any, 0)

	for rows.Next() {
		values := make([]any, len(cols))
		ptrs := make([]any, len(cols))
		for i := range values {
			ptrs[i] = &values[i]
		}
		if err := rows.Scan(ptrs...); err != nil {
			return nil, err
		}
		rowMap := map[string]any{}
		for i, c := range cols {
			rowMap[c] = normalizeDBValue(values[i])
		}
		result = append(result, rowMap)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	return result, nil
}

func normalizeDBValue(v any) any {
	switch t := v.(type) {
	case []byte:
		return string(t)
	default:
		return t
	}
}

func sanitizeColumnName(name string) string {
	return fmt.Sprintf(`"%s"`, strings.ReplaceAll(name, `"`, `""`))
}

func escapeSQL(s string) string {
	return strings.ReplaceAll(s, "'", "''")
}

func TenantTableNameForCache(tenantID string) string {
	return tenant.TableName(tenantID)
}

func segmentListSQL(segments []string) string {
	vals := make([]string, 0, len(segments))
	for _, s := range segments {
		vals = append(vals, fmt.Sprintf("'%s'", escapeSQL(s)))
	}
	return strings.Join(vals, ", ")
}
